File: 07_Basketball/javascript/basketball.js
// BASKETBALL
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var s = [0, 0];
var z;
var d;
var p;
var your_turn;
var game_restart;

function two_minutes()
{
    print("\n");
    print("   *** TWO MINUTES LEFT IN THE GAME ***\n");
    print("\n");
}

function show_scores()
{
    print("SCORE: " + s[1] + " TO " + s[0] + "\n");
}

function score_computer()
{
    s[0] = s[0] + 2;
    show_scores();
}

function score_player()
{
    s[1] = s[1] + 2;
    show_scores();
}

function half_time()
{
    print("\n");
    print("   ***** END OF FIRST HALF *****\n");
    print("SCORE: DARMOUTH: " + s[1] + "  " + os + ": " + s[0] + "\n");
    print("\n");
    print("\n");
}

function foul()
{
    if (Math.random() <= 0.49) {
        print("SHOOTER MAKES BOTH SHOTS.\n");
        s[1 - p] = s[1 - p] + 2;
        show_scores();
    } else if (Math.random() <= 0.75) {
        print("SHOOTER MAKES ONE SHOT AND MISSES ONE.\n");
        s[1 - p] = s[1 - p] + 1;
        show_scores();
    } else {
        print("BOTH SHOTS MISSED.\n");
        show_scores();
    }
}

function player_play()
{
    if (z == 1 || z == 2) {
        t++;
        if (t == 50) {
            half_time();
            game_restart = 1;
            return;
        }
        if (t == 92)
            two_minutes();
        print("JUMP SHOT\n");
        if (Math.random() <= 0.341 * d / 8) {
            print("SHOT IS GOOD.\n");
            score_player();
            return;
        }
        if (Math.random() <= 0.682 * d / 8) {
            print("SHOT IS OFF TARGET.\n");
            if (d / 6 * Math.random() >= 0.45) {
                print("REBOUND TO " + os + "\n");
                return;
            }
            print("DARTMOUTH CONTROLS THE REBOUND.\n");
            if (Math.random() > 0.4) {
                if (d == 6) {
                    if (Math.random() > 0.6) {
                        print("PASS STOLEN BY " + os + " EASY LAYUP.\n");
                        score_computer();
                        return;
                    }
                }
                print("BALL PASSED BACK TO YOU. ");
                your_turn = 1;
                return;
            }
        } else if (Math.random() <= 0.782 * d / 8) {
            print("SHOT IS BLOCKED.  BALL CONTROLLED BY ");
            if (Math.random() <= 0.5) {
                print("DARTMOUTH.\n");
                your_turn = 1;
                return;
            }
            print(os + ".\n");
            return;
        } else if (Math.random() <= 0.843 * d / 8) {
            print("SHOOTER IS FOULED.  TWO SHOTS.\n");
            foul();
            return;
            // In original code but lines 1180-1195 aren't used (maybe replicate from computer's play)
            //        } else if (Math.random() <= 0.9 * d / 8) {
            //            print("PLAYER FOULED, TWO SHOTS.\n");
            //            foul();
            //            return;
        } else {
            print("CHARGING FOUL.  DARTMOUTH LOSES BALL.\n");
            return;
        }
    }
    while (1) {
        if (++t == 50) {
            half_time();
            game_restart = 1;
            return;
        }
        if (t == 92)
            two_minutes();
        if (z == 0) {
            your_turn = 2;
            return;
        }
        if (z <= 3)
            print("LAY UP.\n");
        else
            print("SET SHOT.\n");
        if (7 / d * Math.random() <= 0.4) {
            print("SHOT IS GOOD.  TWO POINTS.\n");
            score_player();
            return;
        }
        if (7 / d * Math.random() <= 0.7) {
            print("SHOT IS OFF THE RIM.\n");
            if (Math.random() <= 2.0 / 3.0) {
                print(os + " CONTROLS THE REBOUND.\n");
                return;
            }
            print("DARMOUTH CONTROLS THE REBOUND.\n");
            if (Math.random() <= 0.4)
                continue;
            print("BALL PASSED BACK TO YOU.\n");
            your_turn = 1;
            return;
        }
        if (7 /d * Math.random() <= 0.875) {
            print("SHOOTER FOULED.  TWO SHOTS.\n");
            foul();
            return;
        }
        if (7 /d * Math.random() <= 0.925) {
            print("SHOT BLOCKED. " + os + "'S BALL.\n");
            return;
        }
        print("CHARGING FOUL.  DARTHMOUTH LOSES THE BALL.\n");
        return;
    }
}

function computer_play()
{
    rebound = 0;
    while (1) {
        p = 1;
        if (++t == 50) {
            half_time();
            game_restart = 1;
            return;
        }
        print("\n");
        z1 = 10 / 4 * Math.random() + 1;
        if (z1 <= 2) {
            print("JUMP SHOT.\n");
            if (8 / d * Math.random() <= 0.35) {
                print("SHOT IS GOOD.\n");
                score_computer();
                return;
            }
            if (8 / d * Math.random() <= 0.75) {
                print("SHOT IS OFF RIM.\n");
                if (d / 6 * Math.random() <= 0.5) {
                    print("DARMOUTH CONTROLS THE REBOUND.\n");
                    return;
                }
                print(os + " CONTROLS THE REBOUND.\n");
                if (d == 6) {
                    if (Math.random() <= 0.75) {
                        print("BALL STOLEN.  EASY LAP UP FOR DARTMOUTH.\n");
                        score_player();
                        continue;
                    }
                    if (Math.random() > 0.6) {
                        print("PASS STOLEN BY " + os + " EASY LAYUP.\n");
                        score_computer();
                        return;
                    }
                    print("BALL PASSED BACK TO YOU. ");
                    return;
                }
                if (Math.random() <= 0.5) {
                    print("PASS BACK TO " + os + " GUARD.\n");
                    continue;
                }
            } else if (8 / d * Math.random() <= 0.90) {
                print("PLAYER FOULED.  TWO SHOTS.\n");
                foul();
                return;
            } else {
                print("OFFENSIVE FOUL.  DARTMOUTH'S BALL.\n");
                return;
            }
        }
        while (1) {
            if (z1 > 3) {
                print("SET SHOT.\n");
            } else {
                print("LAY UP.\n");
            }
            if (7 / d * Math.random() <= 0.413) {
                print("SHOT IS GOOD.\n");
                score_computer();
                return;
            }
            print("SHOT IS MISSED.\n");
            // Spaguetti jump, better to replicate code
            if (d / 6 * Math.random() <= 0.5) {
                print("DARMOUTH CONTROLS THE REBOUND.\n");
                return;
            }
            print(os + " CONTROLS THE REBOUND.\n");
            if (d == 6) {
                if (Math.random() <= 0.75) {
                    print("BALL STOLEN.  EASY LAP UP FOR DARTMOUTH.\n");
                    score_player();
                    break;
                }
                if (Math.random() > 0.6) {
                    print("PASS STOLEN BY " + os + " EASY LAYUP.\n");
                    score_computer();
                    return;
                }
                print("BALL PASSED BACK TO YOU. ");
                return;
            }
            if (Math.random() <= 0.5) {
                print("PASS BACK TO " + os + " GUARD.\n");
                break;
            }
        }
    }
}

// Main program
async function main()
{
    print(tab(31) + "BASKETBALL\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("THIS IS DARTMOUTH COLLEGE BASKETBALL.  YOU WILL BE DARTMOUTH\n");
    print(" CAPTAIN AND PLAYMAKER.  CALL SHOTS AS FOLLOWS:  1. LONG\n");
    print(" (30 FT.) JUMP SHOT; 2. SHORT (15 FT.) JUMP SHOT; 3. LAY\n");
    print(" UP; 4. SET SHOT.\n");
    print("BOTH TEAMS WILL USE THE SAME DEFENSE.  CALL DEFENSE AS\n");
    print("FOLLOWS:  6. PRESS; 6.5 MAN-TO MAN; 7. ZONE; 7.5 NONE.\n");
    print("TO CHANGE DEFENSE, JUST TYPE 0 AS YOUR NEXT SHOT.\n");
    print("YOUR STARTING DEFENSE WILL BE");
    t = 0;
    p = 0;
    d = parseFloat(await input());
    if (d < 6) {
        your_turn = 2;
    } else {
        print("\n");
        print("CHOOSE YOUR OPPONENT");
        os = await input();
        game_restart = 1;
    }
    while (1) {
        if (game_restart) {
            game_restart = 0;
            print("CENTER JUMP\n");
            if (Math.random() > 3.0 / 5.0) {
                print("DARMOUTH CONTROLS THE TAP.\n");
            } else {
                print(os + " CONTROLS THE TAP.\n");
                computer_play();
            }
        }
        if (your_turn == 2) {
            print("YOUR NEW DEFENSIVE ALLIGNMENT IS");
            d = parseFloat(await input());
        }
        print("\n");
        while (1) {
            print("YOUR SHOT");
            z = parseInt(await input());
            p = 0;
            if (z != Math.floor(z) || z < 0 || z > 4)
                print("INCORRECT ANSWER.  RETYPE IT. ");
            else
                break;
        }
        if (Math.random() < 0.5 || t < 100) {
            game_restart = 0;
            your_turn = 0;
            player_play();
            if (game_restart == 0 && your_turn == 0)
                computer_play();
        } else {
            print("\n");
            if (s[1] == s[0]) {
                print("\n");
                print("   ***** END OF SECOND HALF *****\n");
                print("\n");
                print("SCORE AT END OF REGULATION TIME:\n");
                print("        DARTMOUTH: " + s[1] + "  " + os + ": " + s[0] + "\n");
                print("\n");
                print("BEGIN TWO MINUTE OVERTIME PERIOD\n");
                t = 93;
                print("CENTER JUMP\n");
                if (Math.random() > 3.0 / 5.0)
                    print("DARMOUTH CONTROLS THE TAP.\n");
                else
                    print(os + " CONTROLS THE TAP.\n");
            } else {
                print("   ***** END OF GAME *****\n");
                print("FINAL SCORE: DARMOUTH: " + s[1] + "  " + os + ": " + s[0] + "\n");
                break;
            }
        }
    }
}

main();

File: 07_Basketball/java/Basketball.java
import java.lang.Math;
import java.util.*; 
import java.util.Scanner;

/* The basketball class is a computer game that allows you to play as
  Dartmouth College's captain and playmaker
  The game uses set probabilites to simulate outcomes of each posession
  You are able to choose your shot types as well as defensive formations */

public class Basketball {
    int time = 0;
    int[] score = {0, 0};
    double defense = -1;
    List<Double> defense_choices = Arrays.asList(6.0, 6.5, 7.0, 7.5);
    int shot = -1;
    List<Integer> shot_choices = Arrays.asList(0, 1, 2, 3, 4);
    double opponent_chance = 0;
    String opponent = null; 

    public Basketball() {

        // Explains the keyboard inputs
        System.out.println("\t\t\t Basketball");
        System.out.println("\t Creative Computing  Morristown, New Jersey\n\n\n");
        System.out.println("This is Dartmouth College basketball. ");
        System.out.println("Υou will be Dartmouth captain and playmaker.");
        System.out.println("Call shots as follows:");
        System.out.println("1. Long (30ft.) Jump Shot; 2. Short (15 ft.) Jump Shot; "
              + "3. Lay up; 4. Set Shot");
        System.out.println("Both teams will use the same defense. Call Defense as follows:");
        System.out.println("6. Press; 6.5 Man-to-Man; 7. Zone; 7.5 None.");
        System.out.println("To change defense, just type 0 as your next shot.");
        System.out.print("Your starting defense will be? ");

        Scanner scanner = new Scanner(System.in); // creates a scanner

        // takes input for a defense
        if (scanner.hasNextDouble()) {
            defense = scanner.nextDouble();
        }
        else {
            scanner.next();
        }
        
        // makes sure that input is legal
        while (!defense_choices.contains(defense)) {
            System.out.print("Your new defensive allignment is? ");
            if (scanner.hasNextDouble()) {
                defense = scanner.nextDouble();
            }
            else {
                scanner.next();
                continue;
            }
        }

        // takes input for opponent's name
        System.out.print("\nChoose your opponent? ");

        opponent = scanner.next(); 
        start_of_period();
    }

    // adds points to the score
    // team can take 0 or 1, for opponent or Dartmouth, respectively
    private void add_points(int team, int points) {
        score[team] += points;
        print_score();
    }
        

    private void ball_passed_back() {
        System.out.print("Ball passed back to you. ");
        dartmouth_ball();
    }
        
    // change defense, called when the user enters 0 for their shot
    private void change_defense() {
        defense = -1;
        Scanner scanner = new Scanner(System.in); // creates a scanner

        while (!defense_choices.contains(defense)) {
            System.out.println("Your new defensive allignment is? ");
            if (scanner.hasNextDouble()) {
                defense = (double)(scanner.nextDouble()); 
            }
            else {
                continue; 
            }
        }
            
        dartmouth_ball();
    }

    // simulates two foul shots for a player and adds the points
    private void foul_shots(int team) {
        System.out.println("Shooter fouled.  Two shots.");

        if (Math.random() > .49) {
            if (Math.random() > .75) {
                System.out.println("Both shots missed.");
            }
            else {
                System.out.println("Shooter makes one shot and misses one.");
                score[team] += 1; 
            } 
        }
        else {
            System.out.println("Shooter makes both shots.");
            score[team] += 2;
        }
        
        print_score(); 
    }

    // called when time = 50, starts a new period
    private void halftime() {
        System.out.println("\n   ***** End of first half *****\n");
        print_score();
        start_of_period();
    }

    // prints the current score
    private void print_score() {
        System.out.println("Score:  " + score[1] + " to " + score[0] + "\n");
    }

    // simulates a center jump for posession at the beginning of a period
    private void start_of_period() {
        System.out.println("Center jump"); 
        if (Math.random() > .6) {
            System.out.println("Dartmouth controls the tap.\n"); 
            dartmouth_ball();
        }
        else {
            System.out.println(opponent + " controls the tap.\n"); 
            opponent_ball();
        }
    }

    // called when t = 92
    private void two_minute_warning() {
        System.out.println("   *** Two minutes left in the game ***");
    }
        
    // called when the user enters 1 or 2 for their shot
    private void dartmouth_jump_shot() {
        time ++;  
        if (time == 50) {
            halftime();
        }
        else if (time == 92) {
            two_minute_warning();
        }

        System.out.println("Jump Shot.");
        // simulates chances of different possible outcomes
        if (Math.random() > .341 * defense / 8) {
            if (Math.random() > .682 * defense / 8) {
                if (Math.random() > .782 * defense / 8) {
                    if (Math.random() > .843 * defense / 8) {
                        System.out.println("Charging foul. Dartmouth loses ball.\n");
                        opponent_ball();
                    }
                    else {
                        // player is fouled
                        foul_shots(1);
                        opponent_ball();
                    }
                }
                else {
                    if (Math.random() > .5) {
                        System.out.println("Shot is blocked. Ball controlled by " +
                              opponent + ".\n"); 
                        opponent_ball(); 
                    } 
                    else {
                        System.out.println("Shot is blocked. Ball controlled by Dartmouth."); 
                        dartmouth_ball(); 
                    } 
                }
            }
            else {
                System.out.println("Shot is off target."); 
                if (defense / 6 * Math.random() > .45) {
                    System.out.println("Rebound to " + opponent + "\n"); 
                    opponent_ball(); 
                } 
                else {
                    System.out.println("Dartmouth controls the rebound."); 
                    if (Math.random() > .4) {
                        if (defense == 6 && Math.random() > .6) {
                            System.out.println("Pass stolen by " + opponent
                                  + ", easy lay up"); 
                            add_points(0, 2); 
                            dartmouth_ball(); 
                        }    
                        else {
                            // ball is passed back to you
                            ball_passed_back(); 
                        }  
                    }
                    else {
                        System.out.println(""); 
                        dartmouth_non_jump_shot(); 
                    }
                }
            }
        } 
        else {
            System.out.println("Shot is good."); 
            add_points(1, 2); 
            opponent_ball(); 
        }
    }
        
    // called when the user enters 0, 3, or 4
    // lay up, set shot, or defense change
    private void dartmouth_non_jump_shot() {
        time ++; 
        if (time == 50) {
            halftime();
        }
        else if (time == 92) {
            two_minute_warning();
        }
            
        if (shot == 4) {
            System.out.println("Set shot.");
        }
        else if (shot == 3) {
            System.out.println("Lay up."); 
        }
        else if (shot == 0) {
            change_defense(); 
        }
            
        // simulates different outcomes after a lay up or set shot
        if (7/defense*Math.random() > .4) {
            if (7/defense*Math.random() > .7) {
                if (7/defense*Math.random() > .875) {
                    if (7/defense*Math.random() > .925) {
                        System.out.println("Charging foul. Dartmouth loses the ball.\n"); 
                        opponent_ball(); 
                    }
                    else {
                        System.out.println("Shot blocked. " + opponent + "'s ball.\n"); 
                        opponent_ball(); 
                    }
                }
                else {
                    foul_shots(1); 
                    opponent_ball(); 
                }
            }
            else {
                System.out.println("Shot is off the rim."); 
                if (Math.random() > 2/3) {
                    System.out.println("Dartmouth controls the rebound."); 
                    if (Math.random() > .4) {
                        System.out.println("Ball passed back to you.\n"); 
                        dartmouth_ball(); 
                    }
                    else {
                        dartmouth_non_jump_shot(); 
                    }
                }
                else {
                    System.out.println(opponent + " controls the rebound.\n"); 
                    opponent_ball(); 
                }
            }
        }
        else {
            System.out.println("Shot is good. Two points."); 
            add_points(1, 2); 
            opponent_ball(); 
        }
    }
        
    
    // plays out a Dartmouth posession, starting with your choice of shot
    private void dartmouth_ball() {
        Scanner scanner = new Scanner(System.in); // creates a scanner
        System.out.print("Your shot? "); 
        shot = -1; 
        if (scanner.hasNextInt()) {
            shot = scanner.nextInt(); 
        }
        else {
            System.out.println("");
            scanner.next(); 
        }
        
        while (!shot_choices.contains(shot)) {
            System.out.print("Incorrect answer. Retype it. Your shot?");
            if (scanner.hasNextInt()) {
                shot = scanner.nextInt(); 
            }
            else {
                System.out.println("");
                scanner.next(); 
            }
        }

        if (time < 100 || Math.random() < .5) {
            if (shot == 1 || shot == 2) {
                dartmouth_jump_shot();
            }
            else {
                dartmouth_non_jump_shot(); 
            }
        }
        else {
            if (score[0] != score[1]) {
                System.out.println("\n   ***** End Of Game *****"); 
                System.out.println("Final Score: Dartmouth: " + score[1] + "  "
                      + opponent + ": " + score[0]); 
                System.exit(0);
            }
            else {
                System.out.println("\n   ***** End Of Second Half *****"); 
                System.out.println("Score at end of regulation time:"); 
                System.out.println("     Dartmouth: " + score[1] + " " +
                      opponent + ": " + score[0]); 
                System.out.println("Begin two minute overtime period"); 
                time = 93; 
                start_of_period(); 
            }
        }
    }
        
    // simulates the opponents jumpshot
    private void opponent_jumpshot() {
        System.out.println("Jump Shot."); 
        if (8/defense*Math.random() > .35) {
            if (8/defense*Math.random() > .75) {
                if (8/defense*Math.random() > .9) {
                    System.out.println("Offensive foul. Dartmouth's ball.\n"); 
                    dartmouth_ball(); 
                }
                else {
                    foul_shots(0); 
                    dartmouth_ball(); 
                }
            }
            else {
                System.out.println("Shot is off the rim."); 
                if (defense/6*Math.random() > .5) {
                    System.out.println(opponent + " controls the rebound.");
                    if (defense == 6) {
                        if (Math.random() > .75) {
                            System.out.println("Ball stolen. Easy lay up for Dartmouth."); 
                            add_points(1, 2); 
                            opponent_ball(); 
                        }
                        else {
                            if (Math.random() > .5) {
                                System.out.println(""); 
                                opponent_non_jumpshot(); 
                            }
                            else {
                                System.out.println("Pass back to " + opponent +
                                      " guard.\n"); 
                                opponent_ball(); 
                            }
                        }
                    }
                    else {
                        if (Math.random() > .5) {
                            opponent_non_jumpshot(); 
                        }
                        else {
                            System.out.println("Pass back to " + opponent +
                                  " guard.\n"); 
                            opponent_ball(); 
                        }
                    }
                }
                else {
                    System.out.println("Dartmouth controls the rebound.\n"); 
                    dartmouth_ball(); 
                }
            }
        } 
        else {
            System.out.println("Shot is good."); 
            add_points(0, 2); 
            dartmouth_ball(); 
        }
    }

    // simulates opponents lay up or set shot
    private void opponent_non_jumpshot() {
        if (opponent_chance > 3) {
            System.out.println("Set shot."); 
        }
        else {
            System.out.println("Lay up"); 
        }
        if (7/defense*Math.random() > .413) {
            System.out.println("Shot is missed."); 
            if (defense/6*Math.random() > .5) {
                System.out.println(opponent + " controls the rebound."); 
                if (defense == 6) {
                    if (Math.random() > .75) {
                        System.out.println("Ball stolen. Easy lay up for Dartmouth."); 
                        add_points(1, 2); 
                        opponent_ball(); 
                    }
                    else {
                        if (Math.random() > .5) {
                            System.out.println(""); 
                            opponent_non_jumpshot(); 
                        }
                        else {
                            System.out.println("Pass back to " + opponent +
                                  " guard.\n"); 
                            opponent_ball(); 
                        }
                    }
                }
                else {
                    if (Math.random() > .5) {
                        System.out.println(""); 
                        opponent_non_jumpshot(); 
                    }
                    else {
                        System.out.println("Pass back to " + opponent + " guard\n"); 
                        opponent_ball(); 
                    }
                }
            }
            else {
                System.out.println("Dartmouth controls the rebound.\n"); 
                dartmouth_ball(); 
            }
        }
        else {
            System.out.println("Shot is good."); 
            add_points(0, 2); 
            dartmouth_ball(); 
        }
    }

    // simulates an opponents possesion
    // #randomly picks jump shot or lay up / set shot.
    private void opponent_ball() {
        time ++; 
        if (time == 50) {
            halftime();
        }
        opponent_chance = 10/4*Math.random()+1; 
        if (opponent_chance > 2) {
            opponent_non_jumpshot(); 
        }
        else {
            opponent_jumpshot(); 
        }
    }

    public static void main(String[] args) {
        Basketball new_game = new Basketball();
    }
}




